### PR TITLE
Enforce dim check in data feeder

### PR DESCRIPTION
**Pull request by sneaxiy**
This PR is for: 
- Enforce dim check in data feeder 
- Make data feeder support fixed dim inputs

### CODE REVIEW

File: python/paddle/fluid/data_feeder.py [hunk 1]
```diff
@@ -71,10 +71,20 @@ def _feed_impl_(self, data, lod, lod_level):
             for each_data in data:
                 self._feed_impl_(each_data, lod[1:], lod_level - 1)
 
+    def _check_shape_(self, shape):
```
Comment (panyx0718) [inline python/paddle/fluid/data_feeder.py]: no need for `_` suffix.

File: python/paddle/fluid/data_feeder.py [hunk 2]
```diff
@@ -71,10 +71,20 @@ def _feed_impl_(self, data, lod, lod_level):
...
+                    "Shape not match. What is defined in data layer is {}, but receive {}".
+                    format(self.shape, shape))
+
     def done(self):
         arr = numpy.array(self.data, dtype=self.dtype)
-        if self.shape and len(arr.shape) != len(self.shape):
-            arr = arr.reshape(self.shape)
+        if self.shape:
+            if len(arr.shape) != len(self.shape):
+                arr = arr.reshape(self.shape)
```
Comment (panyx0718) [inline python/paddle/fluid/data_feeder.py]: reshape can throw exception. Can catch this and throw a more user-friendly error message.

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/data_feeder.py (hunk 1 above)
Comment (sneaxiy) [reply python/paddle/fluid/data_feeder.py]: Done.

File: python/paddle/fluid/data_feeder.py (hunk 2 above)
Comment (sneaxiy) [reply python/paddle/fluid/data_feeder.py]: Done.

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**